### ISSUE TITLE
A way to convert `code_muni` into any other higher level aggregations

### ISSUE THREAD

**Issue by rafalopespx** [ipeaGIT/geobr#311]
Hello there, 

Is there any rapid and fast way to know from a `code_muni` in which health_region, macro or micro, this municipality is inserted?

This is tricky to do if you trust only on the code, although the `code_muni` carries information from the state level and region, it is not true that part of the `code_muni` can be assigned to a `code_health_region` or `code_health_macrorregion`, just taking the first 5 digits from the `code_muni`

A workaround that I developed is using one of the relational tables from dataSUS to add the `code_health_region` and/or `code_health_macrorregion` to a data.frame with `code_muni`

maybe this can be implemented in `read_municipality`function, as a parameter being passed by the user to return with other code other than `code_muni`

Thanks for the amazing package!

**Comment by rafalopespx** [comment]
Adding on this, at DataSUS's FTP there is a registry of changes of municipalities that were moved from one health region to another, as well as, a table that relates this change and can relate the health regions with the municipalities, by `code_muni` and `code_health_region`

I have the relational database and can send it to anyone who will be working on this conversion, please mail me.

**Comment by rafapereirabr** [comment]
Hi @rafalopespx . Thanks for opening this issue. I like the idea of a table that associates each `code_muni` to the codes of other geographical units. Currently, the closest we have to this is this:

```

df <- geobr::lookup_muni(name_muni = 'all')

head(df)

#>   code_muni             name_muni code_state name_state abbrev_state code_micro
#> 1   1100015 Alta Floresta D'Oeste         11   Rondônia           RO      11006
#> 2   1100023             Ariquemes         11   Rondônia           RO      11003
#> 3   1100031                Cabixi         11   Rondônia           RO      11008
#> 4   1100049                Cacoal         11   Rondônia           RO      11006
#> 5   1100056            Cerejeiras         11   Rondônia           RO      11008
#> 6   1100064     Colorado do Oeste         11   Rondônia           RO      11008
#>          name_micro code_meso         name_meso code_immediate name_immediate
#> 1            Cacoal      1102 Leste Rondoniense         110005         Cacoal
#> 2         Ariquemes      1102 Leste Rondoniense         110002      Ariquemes
#> 3 Colorado do Oeste      1102 Leste Rondoniense         110006        Vilhena
#> 4            Cacoal      1102 Leste Rondoniense         110005         Cacoal
#> 5 Colorado do Oeste      1102 Leste Rondoniense         110006        Vilhena
#> 6 Colorado do Oeste      1102 Leste Rondoniense         110006        Vilhena
#>   code_intermediate name_intermediate
#> 1              1102         Ji-Paraná
#> 2              1101       Porto Velho
#> 3              1102         Ji-Paraná
#> 4              1102         Ji-Paraná
#> 5              1102         Ji-Paraná
#> 6              1102         Ji-Paraná

```

We could probably try to add the code of health regions to this output. Note however, that this output refers to the year 2010. I'm planning to update the function to include the Census 2022 soon.

**Comment by rafapereirabr** [comment]
Hi @rfsaldanha , thanks for the ping. It looks like you are trying to create a correspondence table for each year. Correct?

**Comment by rfsaldanha** [comment]
Yes, to have the corresponding DTB for each year. The problem is that the official IBGE DTB does not agree with the also official IBGE spatial dataset of municipalities of the same year. :-(

**Comment by rafapereirabr** [comment]
That's a problem, indeed. Which one should we trust, the spatial data or the table data?

**Comment by rafalopespx** [comment]
One thing that I encountered too, sometimes the spatial data do not agree with other levels or year of the same spatial data, if we pick all municipalities should this cover the country shapefile map or the state shapefile map, but this does not occur, some years

**Comment by rfsaldanha** [comment]
A "tidy" geobr with topological validation of spatial features would be interesting.

**Comment by rafapereirabr** [comment]
Hi both. Before We we make any data available in **geobr**, we process the data to harmonize column names, projections etc etc and we already "fix" the topolgy by applying `sf::st_make_valid()`. I understand this only fixes topological errors to some extent, though.

For example, here's an example of the problem mentioned by @rafalopespx. The total area of the country polygon is not the same as the sum of areas of each state. This is an inconsistency in the original IBGE data, and there is not much we can do about it. The impression I have is that any attempt to solve this inconsistency should be done by IBGE in the original raw data.

```
library(geobr)
library(sf)
options(scipen = 999)

c <- read_country(year = 2010, simplified = FALSE)
s <- read_state(year = 2010, simplified = FALSE)

area_c <- st_area(c)
area_s <- st_area(s) |> sum()
```

**Comment by rafapereirabr** [comment]
I agree, ideally, we would have all columns added to the output of the lookup table. 

obs. Do you know where to find the list of municipalities in each health region and macro region? I haven't found this table anywhere. It's is possible to determine this using a spatial join operation, but the original boundaries of health regions don't match those of municipalities, which creates some strange results (like a municipality from one state 'included' in the macro health region of another state)

**Comment by rafalopespx** [comment]
I have, I'll send you, where can I send you?

**Comment by rafapereirabr** [comment]
Thanks, @rafalopespx . You can send it to rafa.pereira.br [at] gmail.com.

However, ideally, we would prefer to have an official document or piece of data with this info with an url so we can refer it.

**Comment by rafalopespx** [comment]
Okay, I have found the relational data table in some different services from the Ministry of Health, you can download it directly from here: ftp://ftp.datasus.gov.br/territorio/tabelas/base_territorial.zip, which is the FTP server from DataSUS, and it points to the latest version of it.

Or you can download it from here: https://datasus.saude.gov.br/transferencia-de-arquivos/
and you have to go to Base Territorial at Fonte, then Bases Territoriais at Modelidade, and final Bases Territoriais under the Tipo de Arquivo field

**Comment by rafalopespx** [comment]
If none of them works I can send you the latest one that I downloaded, but I think downloading from there will guarantee you pick the official and latest version of each, and can be easily incorporated into a function

**Comment by JoaoCarabetta** [comment]
Hey, we keep an [updated directory](https://basedosdados.org/dataset/33b49786-fb5f-496f-bb7c-9811c985af8e?table=dffb65ac-9df9-4151-94bf-88c45bfcb056) of codes that a municipality can be related to at Base dos Dados. 

It also brings other codes besides health regions.

you can either download it at the website or R

```
install.packages("basedosdados")
library("basedosdados")

# Defina o seu projeto no Google Cloud
set_billing_id("<YOUR_PROJECT_ID>")

# Para carregar o dado direto no R
query <- bdplyr("br_bd_diretorios_brasil.municipio")
df <- bd_collect(query)```
